diff --git a/compose/ui/ui/src/commonMain/kotlin/androidx/compose/ui/node/LayoutNodeAlignmentLines.kt b/compose/ui/ui/src/commonMain/kotlin/androidx/compose/ui/node/LayoutNodeAlignmentLines.kt
index 295e7fb68221a..1ec0e9defd94d 100644
--- a/compose/ui/ui/src/commonMain/kotlin/androidx/compose/ui/node/LayoutNodeAlignmentLines.kt
+++ b/compose/ui/ui/src/commonMain/kotlin/androidx/compose/ui/node/LayoutNodeAlignmentLines.kt
@@ -146,7 +146,14 @@ internal sealed class AlignmentLines(val alignmentLinesOwner: AlignmentLinesOwne
             if (childOwner.alignmentLines.dirty) {
                 // It did not need relayout, but we still call layout to recalculate
                 // alignment lines.
-                childOwner.layoutChildren()
+                if (alignmentLinesOwner.isPlacingForAlignment) {
+                    val wasPlacingForAlignment = childOwner.isPlacingForAlignment
+                    childOwner.isPlacingForAlignment = true
+                    childOwner.layoutChildren()
+                    childOwner.isPlacingForAlignment = wasPlacingForAlignment
+                } else {
+                    childOwner.layoutChildren()
+                }
             }
             // Add alignment lines on the child node.
             childOwner.alignmentLines.alignmentLineMap.forEach { (childLine, linePosition) ->
diff --git a/compose/ui/ui/src/commonMain/kotlin/androidx/compose/ui/node/LayoutNodeLayoutDelegate.kt b/compose/ui/ui/src/commonMain/kotlin/androidx/compose/ui/node/LayoutNodeLayoutDelegate.kt
index 6597a1bd02d7a..5ef592dc06ea3 100644
--- a/compose/ui/ui/src/commonMain/kotlin/androidx/compose/ui/node/LayoutNodeLayoutDelegate.kt
+++ b/compose/ui/ui/src/commonMain/kotlin/androidx/compose/ui/node/LayoutNodeLayoutDelegate.kt
@@ -414,6 +414,12 @@ internal const val MeasuredTwiceErrorMessage: String =
  */
 internal interface AlignmentLinesOwner : Measurable {
 
+    /**
+     * Whether the [AlignmentLinesOwner] is currently laying out its children only for the purposes
+     * of calculating the alignment lines. See [LookaheadCapablePlaceable.isPlacingForAlignment].
+     */
+    var isPlacingForAlignment: Boolean
+
     val placeOrder: Int
 
     /** InnerNodeCoordinator of the LayoutNode that the AlignmentLinesOwner operates on. */
diff --git a/compose/ui/ui/src/commonMain/kotlin/androidx/compose/ui/node/LookaheadPassDelegate.kt b/compose/ui/ui/src/commonMain/kotlin/androidx/compose/ui/node/LookaheadPassDelegate.kt
index f8e411ecffb8a..0b93e10ae956d 100644
--- a/compose/ui/ui/src/commonMain/kotlin/androidx/compose/ui/node/LookaheadPassDelegate.kt
+++ b/compose/ui/ui/src/commonMain/kotlin/androidx/compose/ui/node/LookaheadPassDelegate.kt
@@ -358,6 +358,12 @@ internal class LookaheadPassDelegate(
         return alignmentLines.getLastCalculation()
     }
 
+    override var isPlacingForAlignment: Boolean
+        get() = innerCoordinator.lookaheadDelegate?.isPlacingForAlignment == true
+        set(value) {
+            innerCoordinator.lookaheadDelegate?.isPlacingForAlignment = value
+        }
+
     override val parentAlignmentLinesOwner: AlignmentLinesOwner?
         get() = layoutNode.parent?.layoutDelegate?.lookaheadAlignmentLinesOwner
 
diff --git a/compose/ui/ui/src/commonMain/kotlin/androidx/compose/ui/node/MeasurePassDelegate.kt b/compose/ui/ui/src/commonMain/kotlin/androidx/compose/ui/node/MeasurePassDelegate.kt
index 57a57d334651c..a97bda01a352a 100644
--- a/compose/ui/ui/src/commonMain/kotlin/androidx/compose/ui/node/MeasurePassDelegate.kt
+++ b/compose/ui/ui/src/commonMain/kotlin/androidx/compose/ui/node/MeasurePassDelegate.kt
@@ -843,6 +843,12 @@ internal class MeasurePassDelegate(private val layoutNodeLayoutDelegate: LayoutN
         return alignmentLines.getLastCalculation()
     }
 
+    override var isPlacingForAlignment: Boolean
+        get() = innerCoordinator.isPlacingForAlignment
+        set(value) {
+            innerCoordinator.isPlacingForAlignment = value
+        }
+
     override val parentAlignmentLinesOwner: AlignmentLinesOwner?
         get() = layoutNode.parent?.layoutDelegate?.alignmentLinesOwner
 
